source/threads.tex

```diff
@@ -3566,6 +3568,652 @@
+where each phase allows at most an expected number of threads to block
+until the expected number of threads arrive at the barrier.
+\begin{note}
+A barrier is useful for managing repeated tasks
+that are handled by multiple threads.
+\end{note}
+
+\rSec3[barrier.syn]{Header \tcode{<barrier>} synopsis}
+\indexhdr{barrier}%
+
+\begin{codeblock}
+namespace std {
+  template<class CompletionFunction = @\seebelow@>
+    class barrier;
+}
+\end{codeblock}
+
+\rSec3[thread.barrier.class]{Class template \tcode{barrier}}
+
+\begin{codeblock}
+namespace std {
+  template<class CompletionFunction = @\seebelow@>
+  class barrier {
+  public:
+    using arrival_token = see below;
+
+    constexpr explicit barrier(ptrdiff_t phase_count,
+                               CompletionFunction f = CompletionFunction());
+    ~barrier();
+
+    barrier(const barrier&) = delete;
+    barrier& operator=(const barrier&) = delete;
+
+    [[nodiscard]] arrival_token arrive(ptrdiff_t update = 1);
+    void wait(arrival_token&& arrival) const;
+
+    void arrive_and_wait();
+    void arrive_and_drop();
+
+  private:
+    CompletionFunction completion;      // \expos
+  };
+}
+\end{codeblock}
+
+\pnum
+Each \defn{barrier phase} consists of the following steps:
+\begin{itemize}
+\item
+  The expected count is decremented
+  by each call to \tcode{arrive} or \tcode{arrive_and_drop}.
+\item
+  When the expected count reaches zero, the phase completion step is run.
+  For the specialization
+  with the default value of the \tcode{CompletionFunction} template parameter,
+  the completion step is run
+  as part of the call to \tcode{arrive} or \tcode{arrive_and_drop}
+  that caused the expected count to reach zero.
+  For other specializations,
+  the completion step is run on one of the threads
+  that arrived at the barrier during the phase.
+\item
+  When the completion step finishes,
+  the expected count is reset
+  to what was specified by the \tcode{expected} argument to the constructor,
+  possibly adjusted by calls to \tcode{arrive_and_drop}, and
+  the next phase starts.
+\end{itemize}
+
+\pnum
+Each phase defines a \defn{phase synchronization point}.
+Threads that arrive at the barrier during the phase
+can block on the phase synchronization point by calling\tcode{wait}, and
+will remain blocked until the phase completion step is run.
+
+\pnum
+The \defn{phase completion step}
+that is executed at the end of each phase has the following effects:
+\begin{itemize}
+\item Invokes the completion function, equivalent to \tcode{completion()}.
+\item Unblocks all threads that are blocked on the phase synchronization point.
+\end{itemize}
+The end of the completion step strongly happens before
+the returns from all calls that were unblocked by the completion step.
+For specializations that do not have
+the default value of the \tcode{CompletionFunction} template parameter,
+the behavior is undefined if any of the barrier object’s member functions
```

Comment:
"object's" rendered as "objectâĂŹs".  Check all uses of quotes in these papers.

Reply:
Done, and fixed one more.